source/common/http/filter/fault_filter.cc

```diff
@@ -0,0 +1,159 @@
+                                         header_map.getString("value", EMPTY_STRING));
+    }
+  }
+}
+
+FaultFilter::FaultFilter(FaultFilterConfigPtr config) : config_(config) {}
+
+FaultFilter::~FaultFilter() { ASSERT(!delay_timer_); }
+
+// Delays and aborts are independent events. One can inject a delay
+// followed by an abort or inject just a delay or abort. In this callback,
+// if we inject a delay, then we will inject the abort in the delay timer
+// callback.
+FilterHeadersStatus FaultFilter::decodeHeaders(HeaderMap& headers, bool) {
+  // Check for header matches first
+  if (!Router::ConfigUtility::matchHeaders(headers, config_->filterHeaders())) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  if (config_->runtime().snapshot().featureEnabled("fault.http.delay.fixed_delay_percent",
+                                                   config_->delayPercent())) {
+    uint64_t duration_ms = config_->runtime().snapshot().getInteger(
+        "fault.http.delay.fixed_duration_ms", config_->delayDuration());
+
+    // Delay only if the duration is >0ms
+    if (0 != duration_ms) {
+      delay_timer_ =
+          callbacks_->dispatcher().createTimer([this]() -> void { postDelayInjection(); });
+      delay_timer_->enableTimer(std::chrono::milliseconds(duration_ms));
+      config_->stats().delays_injected_.inc();
+      return FilterHeadersStatus::StopIteration;
+    }
+  }
+
+  if (config_->runtime().snapshot().featureEnabled("fault.http.abort.abort_percent",
+                                                   config_->abortPercent())) {
+    // todo check http status codes obtained from runtime
+    Http::HeaderMapPtr response_headers{new HeaderMapImpl{
+        {Headers::get().Status, std::to_string(config_->runtime().snapshot().getInteger(
+                                    "fault.http.abort.http_status", config_->abortCode()))}}};
+    callbacks_->encodeHeaders(std::move(response_headers), true);
+    config_->stats().aborts_injected_.inc();
+    return FilterHeadersStatus::StopIteration;
+  }
+
+  return FilterHeadersStatus::Continue;
+}
+
+FilterDataStatus FaultFilter::decodeData(Buffer::Instance&, bool) {
+  return FilterDataStatus::Continue;
+}
+
+FilterTrailersStatus FaultFilter::decodeTrailers(HeaderMap&) {
+  return FilterTrailersStatus::Continue;
+}
+
+FaultFilterStats FaultFilter::generateStats(const std::string& prefix, Stats::Store& store) {
+  std::string final_prefix = prefix + "fault.";
+  return {ALL_FAULT_FILTER_STATS(POOL_COUNTER_PREFIX(store, final_prefix))};
+}
+
+void FaultFilter::onResetStream() { resetTimerState(); }
+
+void FaultFilter::postDelayInjection() {
+  resetTimerState();
+  // Delays can be followed by aborts
+  if (config_->runtime().snapshot().featureEnabled("fault.http.abort.abort_percent",
+                                                   config_->abortPercent())) {
+    Http::HeaderMapPtr response_headers{new HeaderMapImpl{
```

Comment:
Can we have a helper function that sends the response that is called from both here as well as decodeHeaders() so we avoid the code duplication.